src/web/hooks/useTiming.js

```diff
@@ -0,0 +1,109 @@
+/* SPDX-FileCopyrightText: 2024 Greenbone AG
+ *
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+import {useEffect, useCallback, useState} from 'react';
+
+import logger from 'gmp/log';
+
+import {hasValue, isFunction, isDefined} from 'gmp/utils/identity';
+
+import useInstanceVariable from './useInstanceVariable';
+
+const log = logger.getLogger('web.hooks.useTiming');
+
+/**
+ * Hook to start a timer that calls a function after a given timeout
+ *
+ * @param {Function} doFunc The function to call
+ * @param {Number|Function} timeout The timeout in milliseconds or a function that returns the timeout
+ * @returns Array of startTimer function, clearTimer function and boolean isRunning
+ */
+const useTiming = (doFunc, timeout) => {
+  const timer = useInstanceVariable({});
+  const [timerId, setTimerId] = useState(); // store timerId in state too to trigger re-render if it changes
+  const isRunning = !!timerId;
+  timer.doFunc = doFunc; // always use the latest version of the function
+
+  const updateTimerId = useCallback(
+    newTimerId => {
+      timer.timerId = newTimerId;
+      setTimerId(newTimerId);
+    },
+    [timer],
+  );
+
+  const startTimer = useCallback(() => {
+    if (hasValue(timer.timerId)) {
+      log.debug('Not starting timer. A timer is already running.', {
+        timer: timer.timerId,
+      });
+      return;
+    }
+
+    const timeoutValue = isFunction(timeout) ? timeout() : timeout;
+
+    if (!hasValue(timeoutValue) || timeoutValue < 0) {
+      log.debug('Not starting timer because timeout value was', timeoutValue);
+      return;
+    }
+
+    updateTimerId(
+      setTimeout(() => {
+        log.debug('Timer with id', timer.timerId, 'fired.');
+
+        const promise = timer.doFunc();
+
+        if (isDefined(promise?.then)) {
+          promise
+            .then(() => {
+              updateTimerId(undefined);
+              timer.startTimer();
+            })
+            .catch(() => {
+              updateTimerId(undefined);
+            });
+        } else {
+          updateTimerId(undefined);
+        }
+      }, timeoutValue),
+    );
```

Comment:
Maybe `try/catch` and `async/await` are a bit more readable. Something like this?

```suggestion
    updateTimerId(
      setTimeout(async () => {
        log.debug('Timer with id', timer.timerId, 'fired.');
    
        try {
          const promise = await timer.doFunc();
          updateTimerId(undefined);
          if (promise) {
            timer.startTimer();
          }
        } catch  {
          updateTimerId(undefined);
        }
      }, timeoutValue),
    );
```

Reply:
Currently it is not necessary that `timer.doFunc` is a promise. It can be a normal function. Therefore I am using ` if isDefined(promise?.then)` to check whether it is a promise or not. Maybe something like this is possible
```js
        try {
          const promise = timer.doFunc();
          updateTimerId(undefined);
          if (promise?.then) {
            await promise;
            timer.startTimer();
          }
        } catch  {
          updateTimerId(undefined);
        }
      }, timeoutValue),
```
?

Reply:
Ok, yeah, got it. Looks already much more readable with the try/catch :rocket: